pkg/operator/minio/controller.go

```diff
@@ -0,0 +1,301 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package to manage a Minio object store.
+package minio
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/coreos/pkg/capnslog"
+	opkit "github.com/rook/operator-kit"
+	miniov1alpha1 "github.com/rook/rook/pkg/apis/minio.rook.io/v1alpha1"
+	"github.com/rook/rook/pkg/clusterd"
+	"k8s.io/api/apps/v1beta2"
+	"k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+// TODO: A lot of these constants are specific to the KubeCon demo. Let's
+// revisit these and determine what should be specified in the resource spec.
+const (
+	customResourceName       = "objectstore"
+	customResourceNamePlural = "objectstores"
+	minioPVCName             = "minio-pvc"
+	minioLabel               = "minio"
+	minioServerSuffixFmt     = "%s.svc.cluster.local"
+	minioVolumeName          = "data"
+	minioCtrName             = "minio"
+	minioCtrImage            = "minio/minio:RELEASE.2018-04-19T22-54-58Z"
```

Comment:
we are building the `rook/minio` image based off this upstream minio image. it seems like this controller should only use the `rook/minio` image that the operator tells it to (currently being passed into this controller as `rookImage`).  This keeps the versions of the operator and deployed object stores in sync.  the upstream minio image shouldn't be running directly at all.

Reply:
Can you elaborate on why? Currently, we're taking the upstream image and just copying a Rook binary into it for rook/minio:master.

Reply:
Also, the entrypoint for the rookImage is Rook. I don't see a reason to spawn a Minio process from within Rook unless we want the Rook PID to be the parent for some reason? Perhaps I'm missing something though, so I'll stand by for a response before making a change.

Reply:
The reason to use the same image as the operator is so that the cluster is always in sync with the version of the operator.  The user specifies what version of the operator they want to use in operator.yaml and then in this controller we are just hard coding a version of minio that may not match what the operator's image is based on.  Also, when we want to update the version of minio in the future, we'd have to update it in the image build scripts and then also update it here in the code too.  I think it's a better approach to use the same image that the operator uses so it honors the version the user requested and so that operator and cluster have matching versions.

The entrypoint being rook by default doesn't matter, it can be set to anything using the `v1.Container.Command` field. The rook binary wouldn't come into the picture at all, the container would launch directly into the minio binary instead.  See the comments on https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2014 for guidance.